examples/decoding/cohy_decomposition.py

```diff
@@ -0,0 +1,532 @@
+# ``spectral_connectivity_...()`` functions: the transformed data; the indices
+# returned from :meth:`~mne_connectivity.decoding.CaCoh.get_transformed_indices`; and
+# the corresponding bivariate method (``"coh"`` and ``"cohy"`` for CaCoh; ``"imcoh"``
+# for MIC).
+#
+# As you can see, the connectivity profile of the transformed data using filters fit on
+# the first 30 epochs is very similar to the connectivity profile when using filters fit
+# on the last 30 epochs. This shows that the filters are generalisable, able to extract
+# the same components of connectivity which they were trained on from new data.
+
+# %%
+
+# Connectivity profile of last 30 epochs (filters fit to these epochs)
+con_cacoh_last = spectral_connectivity_epochs(
+    epochs.get_data(item=np.arange(N_EPOCHS // 2, N_EPOCHS)),
+    method="cacoh",
+    indices=([indices[0]], [indices[1]]),
+    fmin=5,
+    fmax=35,
+    sfreq=epochs.info["sfreq"],
+    rank=([3], [3]),
+)
+ax = plt.subplot(111)
+ax.plot(
+    con_cacoh_last.freqs,
+    np.abs(con_cacoh_last.get_data()[0]),
+    label="CaCoh (filters trained\non epochs 30-60)",
+)
+
+# Connectivity profile of last 30 epochs (no filters)
+con_coh_last = spectral_connectivity_epochs(
+    epochs.get_data(item=np.arange(N_EPOCHS // 2, N_EPOCHS)),
+    method="coh",
+    indices=seed_target_indices(indices[0], indices[1]),
+    fmin=5,
+    fmax=35,
+    sfreq=epochs.info["sfreq"],
+)
+ax.plot(
+    con_coh_last.freqs, np.mean(np.abs(con_coh_last.get_data()), axis=0), label="Coh"
+)
+
+# Connectivity profile of last 30 epochs (filters fit to first 30 epochs)
+con_cacoh_last_from_first = spectral_connectivity_epochs(
+    epochs_transformed,
+    method="coh",
+    indices=indices_transformed,
+    fmin=5,
+    fmax=35,
+    sfreq=epochs.info["sfreq"],
+)
+ax.plot(
+    con_cacoh_last_from_first.freqs,
+    np.abs(con_cacoh_last_from_first.get_data()[0]),
+    label="CaCoh (filters trained\non epochs 0-30)",
+)
+ax.axvspan(FMIN, FMAX, color="grey", alpha=0.2, label="Fitted freq. band")
+ax.set_xlabel("Frequency (Hz)")
+ax.set_ylabel("Connectivity (A.U.)")
+ax.set_title("Epochs 30-60")
+plt.legend()
+plt.show()
+
+########################################################################################
+# In addition to assessing the validity of the approach, we can also look at the time
+# taken to run the analysis. Below we present a scenario resembling an online sliding
+# window approach typical of a BCI system. We consider the first 30 epochs to be the
+# training data that the filters should be fit to, and the last 30 epochs to be the
+# windows of data that the filters should be applied to, transforming and computing the
+# connectivity of each window (epoch) of data sequentially.
+#
+# Doing so, we see that once the filters have been fit, it takes only a few milliseconds
+# to transform each window of data and compute its connectivity.
```

Comment:
In addition to it being quicker, would the traditional argument of overfitting also be valid here? If you compute the filters on a single epoch, the filters will be super-tuned to that specific epoch. The data is simulated with coherence in the 15-20 Hz frequency band. With this in mind, you can see in the plot:

- filters fitted on the first 30 epochs and then applied to the final 30 epochs nicely capture the coherence in the 15-20 Hz band and not much outside of that frequency band. (proper cross-validation)
- filters fitted on 30 epochs and then applied to the same 30 epochs also capture some coherence outside of the 15-20 Hz band

I would imagine fitting the filters to a single epoch and then applying them to that same epoch will show even more coherence outside the 15-20 Hz band, even though there should not be any coherence there given our simulation.

Reply:
> In addition to it being quicker, would the traditional argument of overfitting also be valid here? If you compute the filters on a single epoch, the filters will be super-tuned to that specific epoch.

Yeah, that's a very good point. Will add this.